be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/RollCallHandler.scala

```diff
@@ -40,10 +40,41 @@ class RollCallHandler(dbRef: => AskableActorRef) extends MessageHandler {
   private val unknownAnswer: String = "Database actor returned an unknown answer"
 
   def handleCreateRollCall(rpcRequest: JsonRpcRequest): GraphMessage = {
-    val ask: Future[GraphMessage] = dbAskWritePropagate(rpcRequest)
-    Await.result(ask, duration)
+    rpcRequest.getParamsMessage match {
+      case Some(message: Message) =>
+        val data: CreateRollCall = message.decodedData.get.asInstanceOf[CreateRollCall]
+        // we are using the rollcall id instead of the message_id at rollcall creation
+        val rollCallChannel: Channel = Channel(s"${Channel.ROOT_CHANNEL_PREFIX}${data.id}")
+        val ask =
+          for (
+            channelExists <- dbActor ? DbActor.ChannelExists(rollCallChannel)
+          ) yield channelExists match {
+            case DbActor.DbActorAck =>
+              Future(Right(PipelineError(0, "", rpcRequest.getId)))
+            case _ =>
+              for (
+                _ <- dbActor ? DbActor.CreateChannel(rollCallChannel, ObjectType.ROLL_CALL);
+                x <- dbAskWritePropagate(rpcRequest)
+              )
+              yield x
+          }
+
+        Await.result(Await.result(ask, duration), duration) match {
+          case success@Left(_) =>
+            success
+          case error@Right(_) => error
+          case _ => Right(
+            PipelineError(ErrorCodes.ALREADY_EXISTS.id, "Rollcall already exists in db", rpcRequest.getId)
+          )
+        }
```

Comment:
Turn double `for` comprehensions into a single one using the fact that:
  1. `future2` will ONLY get executed iff `future1` is a success (i.e. has terminated and is successful)
  2.  In this example, if `future1` is of type `A`, then you can match (IF NEEDED ONLY) on the value `a: A` (value contained in `future1`) using the matcher `Some(Success(a))`
  3.  If any of the futures fails (i.e. terminates and failure), then the matching value is of type `Option[Failure[DbActorNAckException]]` with a defined `Option`
  4.  if any of the futures times out, then the matching value is of type `Option[...]` that is empty, i.e. `None`

```scala
val ask = for {
  future1 <- dbActor ? DbActor.Something()
  future2 <- dbActor ? DbActor.SomethingElse()
} yield (future1)

Await.ready(ask, timeout).value match { case _ => /* ... */ }
```